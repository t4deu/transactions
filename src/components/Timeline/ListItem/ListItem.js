// @flow
import React, { PureComponent } from 'react';
import { Text, View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import currency from 'currency.js';

import type { Transaction } from '../../../state/ducks/transactions/types';

import styles from './ListItem.styles';

type Props = {
  transaction: Transaction,
};

export default class ListItem extends PureComponent<Props> {
  renderTime(transaction:Transaction) {
    return (
      <View style={styles.timeWrapper}>
        <Text style={styles.time}>{transaction.createdAt}</Text>
      </View>
    );
  }

  renderDetails(transaction:Transaction) {
    const detailStyle = this.props.last ? styles.lastDetail : styles.defaultDetail;

    return (
      <View style={[styles.details, detailStyle]}>
        <View style={styles.detail}>{this.renderDescription(transaction)}</View>
        <View style={styles.separator} />
      </View>
    );
  }

  renderDescription(transaction:Transaction) {
    return (
      <View>
        {this.renderHeader(transaction)}
        <Text style={styles.description}>{transaction.description}</Text>
        <Text style={styles.price}>R$ {currency(transaction.amount).format()}</Text>
      </View>
    );
  }

  renderHeader(transaction:Transaction) {
    const title = transaction.type === 'credit' ? 'Crédito' : 'Débito';
    return (
      <View style={styles.header}>
        <Text style={styles.title}>{title}</Text>
        <View style={styles.timeWrapper}>
          <Text style={styles.time}>{transaction.createdAt}</Text>
        </View>
      </View>
    );
  }

  renderCircle(transaction:Transaction) {
    return transaction.type === 'credit' ? (
      <View style={styles.circle}>
        <Ionicons name="ios-checkmark-circle" size={32} color="#9bd533" />
      </View>
    ) : (
      <View style={styles.circle}>
        <Ionicons name="ios-alert" size={32} color="#e6605e" />
      </View>
    );
  }

  render() {
    const { transaction } = this.props;

    return (
      <View style={styles.container}>
        {this.renderDetails(transaction)}
        {this.renderCircle(transaction)}
      </View>
    );
  }
}
