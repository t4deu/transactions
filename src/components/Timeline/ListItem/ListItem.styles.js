export default {
  container: {
    flexDirection: 'row',
    flex: 1,
    justifyContent: 'center',
  },
  header: {
    flexDirection: 'row',
  },
  timeWrapper: {
    alignItems: 'flex-end',
    flex: 1,
  },
  time: {
    textAlign: 'right',
    color: '#858585',
  },
  details: {
    borderRightWidth: 0,
    marginLeft: 20,
    paddingLeft: 25,
    paddingRight: 20,
    borderLeftWidth: 2,
    flexDirection: 'column',
    flex: 1,
  },
  defaultDetail: {
    borderColor: '#3c2b3e',
  },
  lastDetail: {
    borderColor: 'rgba(0,0,0,0)',
  },
  separator: {
    height: 1,
    backgroundColor: '#aaa',
    marginTop: 10,
    marginBottom: 10,
  },
  title: {
    fontSize: 12,
    fontWeight: 'bold',
  },
  detail: {
    paddingTop: 4,
    paddingBottom: 10,
  },
  description: {
    marginTop: 10,
  },
  price: {
    fontSize: 18,
  },
  circle: {
    borderRadius: 10,
    position: 'absolute',
    left: 8,
    alignItems: 'center',
    backgroundColor: '#fff',
    justifyContent: 'center',
  },
};
