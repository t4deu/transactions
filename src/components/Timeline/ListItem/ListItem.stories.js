import React from 'react';
import { storiesOf } from '@storybook/react-native';

import ListItem from './ListItem';

storiesOf('Timeline item', module)
  .add('with Credit', () => (
    <ListItem
      transaction={{
        id: 1,
        type: 'credit',
        amount: 10.5,
        description: 'Depósito poupança',
        createdAt: '2018-03-20',
      }}
    />
  ))
  .add('with Debit', () => (
    <ListItem
      transaction={{
        id: 1,
        type: 'debit',
        amount: 50.9,
        description: 'Confraria da Pizza',
        createdAt: '2018-03-20',
      }}
    />
  ))
  .add('is last item', () => (
    <ListItem
      last
      transaction={{
        id: 1,
        type: 'credit',
        amount: 10.5,
        description: 'Depósito poupança',
        createdAt: '2018-03-20',
      }}
    />
  ));;
