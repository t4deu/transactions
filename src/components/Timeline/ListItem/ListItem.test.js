import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ListItem from './ListItem';

describe('<ListItem>', () => {
  it('renders without issues', () => {
    const transaction = {
      id: 1,
      type: 'credit',
      amount: 10,
      description: 'he Beginner Archery and Beginner Crossbow',
      createdAt: '10:30',
    };
    const component = shallow(<ListItem transaction={transaction} />);

    expect(toJson(component)).toMatchSnapshot();
  });
});
