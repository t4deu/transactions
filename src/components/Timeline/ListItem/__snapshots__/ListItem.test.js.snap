// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem> renders without issues 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderLeftWidth": 2,
          "borderRightWidth": 0,
          "flex": 1,
          "flexDirection": "column",
          "marginLeft": 20,
          "paddingLeft": 25,
          "paddingRight": 20,
        },
        Object {
          "borderColor": "#3c2b3e",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "paddingBottom": 10,
          "paddingTop": 4,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 12,
                "fontWeight": "bold",
              }
            }
          >
            Cr√©dito
          </Text>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 1,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#858585",
                  "textAlign": "right",
                }
              }
            >
              20/03/18
            </Text>
          </View>
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          he Beginner Archery and Beginner Crossbow
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 18,
            }
          }
        >
          R$ 10,00
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#aaa",
          "height": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 10,
        "justifyContent": "center",
        "left": 8,
        "position": "absolute",
      }
    }
  >
    <Icon
      allowFontScaling={false}
      color="#9bd533"
      name="ios-checkmark-circle"
      size={32}
    />
  </View>
</View>
`;
